src/compiler/parser.ts

```diff
@@ -522,6 +522,76 @@ namespace ts {
         }
     }
 
+    /** @internal */
+    /**
+     * Invokes a callback for each child of the given node. The 'cbNode' callback is invoked for all child nodes
+     * stored in properties. If a 'cbNodes' callback is specified, it is invoked for embedded arrays; additionally,
+     * unlike `forEachChild`, embedded arrays are flattened and the 'cbNode' callback is invoked for each element.
+     *  If a callback returns a truthy value, iteration stops and that value is returned. Otherwise, undefined is returned.
+     *
+     * @param node a given node to visit its children
+     * @param cbNode a callback to be invoked for all child nodes
+     * @param cbNodes a callback to be invoked for embedded array
+     *
+     * @remarks Unlike `forEachChild`, `forEachChildRecursively` handles recursively invoking the traversal on each child node found,
+     * and while doing so, handles traversing the structure without relying on the callstack to encode the tree structure.
+     */
+    export function forEachChildRecursively<T>(rootNode: Node, cbNode: (node: Node, parent: Node) => T | "skip" | undefined, cbNodes?: (nodes: NodeArray<Node>, parent: Node) => T | "skip" | undefined): T | undefined {
+
+        const stack: Node[] = [rootNode];
+        while (stack.length) {
+            const parent = stack.pop()!;
+            const res = visitAllPossibleChildren(parent, gatherPossibleChildren(parent));
+            if (res) {
+                return res;
+            }
+        }
+
+        return;
+
+        function gatherPossibleChildren(node: Node) {
+            const children: (Node | NodeArray<Node>)[] = [];
+            forEachChild(node, addWorkItem, addWorkItem); // By using a stack above and `unshift` here, we emulate a depth-first preorder traversal
+            return children;
+
+            function addWorkItem(n: Node | NodeArray<Node>) {
+                children.unshift(n);
+            }
+        }
+
+        function visitAllPossibleChildren(parent: Node, children: readonly (Node | NodeArray<Node>)[]) {
+            for (const child of children) {

```

Comment:
is it intended, that children are visited in reverse order?

Reply:
In the `for` loop below? IIRC, yes; because a stack is in use as the work-tracking structure, when we add things to the stack, we want to add them in the opposite order we want them to get popped off in.

Reply:
I got that it's necessary for the stack order, but adding to the stack and calling `cbNode` is done in the same iteration. Therefore children of a given parent are visited in reversed order.

Given the tree from the comment above the current logic is as follows:

* gather children of 1 in reverse order: [3, 2]
* visitNode 3, add 3 to the stack
* visitNode 2, add 2 to the stack
* pop 2 from the stack
* recurse into 2
  * gather children of 2 in reverse order: [5, 4]
  * visitNode 5, add to stack
  * visitNode 4, add to stack
  * pop 4 from the stack
  * pop 5 from the stack
* pop 3 from the stack
  * ... 

This doesn't match the inline comment about this being "depth-first preorder".